CIP-0133/README.md

```diff
@@ -0,0 +1,116 @@
+
+We also did preliminary [benchmarks](https://github.com/dkaidalov/bench-blst-msm/) to compare an [optimized blst](https://github.com/supranational/blst/blob/e99f7db0db413e2efefcfd077a4e335766f39c27/src/multi_scalar.c) implementation of MSM with naive implementation using just blst group operations. We used Rust bindings to do these benchmarks (the underlying bindings are the same as used in [cardano-base for bslt](https://github.com/IntersectMBO/cardano-base/blob/master/cardano-crypto-class/src/Cardano/Crypto/EllipticCurve/BLS12_381/Internal.hs#L353), which means we can expect similar behavior for the Haskell stack). The results for G1 group are the following (more results in the [repository](https://github.com/dkaidalov/bench-blst-msm/)):
+
+| MSM Size | MSM Optimized Average Time | MSM Naive Average Time | Ratio Naive/Optimized |
+|---------:|---------------------------:|-----------------------:|----------------------:|
+|       15 |                  728.565µs |                1.263ms |                  1.73 |
+|       20 |                  816.508µs |                1.499ms |                  1.83 |
+|       25 |                  956.735µs |                1.837ms |                  1.92 |
+|       30 |                    1.022ms |                2.055ms |                  2.01 |
+|       35 |                 428.765µs* |                2.715ms |                 6.13* |
+|       40 |                   469.49µs |                2.882ms |                  6.33 |
+|       45 |                  445.014µs |                3.288ms |                  7.37 |
+|       50 |                  533.669µs |                3.933ms |                  7.38 |
+|      100 |                   770.71µs |                7.307ms |                  9.48 |
+|      200 |                    1.065ms |                14.07ms |                  13.2 |
+|      300 |                    1.177ms |               21.605ms |                 18.34 |
+|      400 |                    1.312ms |               27.405ms |                 20.87 |
+|     1000 |                    2.584ms |               65.725ms |                 25.43 |
+|     2000 |                    4.108ms |              131.453ms |                 31.99 |
+|     3000 |                    5.510ms |              195.986ms |                 35.56 |
+|     4000 |                     7.00ms |              263.722ms |                 37.67 |
+
+\* _the sudden time improvement after 32 points is attributed to the inner workings of the blst library, which may operate differently for the MSM of size less than 32 (this should be carefully analyzed when establishing costing function for MSM built-in)._
```

Comment:
Good question, I think that in practical application you can choose which group to use (as you can move things between the two groups naturally). And due to G2 being bigger and pricier, the default probably will be G1. But I do not exclude possible cases where both are handy.

Reply:
I believe the sudden time improvement after 32 points is due to [this](https://github.com/supranational/blst/blob/master/bindings/rust/src/pippenger.rs#L165) condition.
For some reason Rust bindings basically do a naive MSM calculation for <32 points trying instead to parallelize. Curiously, for non-parallelizable version [here](https://github.com/supranational/blst/blob/master/bindings/rust/src/pippenger-no_std.rs#L78) they don't do this trick. But I didn't found argumentation for this, and I'm not sure it would be beneficial to do the same in Haskell.

Reply:
> I believe the sudden time improvement after 32 points is due to [this](https://github.com/supranational/blst/blob/master/bindings/rust/src/pippenger.rs#L165) condition.

Right, that's in the Rust code and it looks as if it's doing stuff in parallel in some cases.  I think we probably don't want to try doing anything like that in the Plutus implementation.

Reply:
Did your benchmarks actually use a rust implementation that does some of the MSM calculations in parallel?  I took a look at the code but I didn't follow it into the innards of the implementation.  If there was some parallelism involved then the speedups may be a bit misleading, but I guess we'll find out when we can experiment with a Plutus version.  

FYI I checked and the node is currently run with `-N2` and I think it's very unlikely that we'd be allowed to use a library function that has a significant amount of parallelism because there's no guarantee that all of the nodes in the network would benefit equally from that: some nodes might not be able to process scripts that other ones have no problem with.

Reply:
yes, our benchmarks use rust bindings that utilize parallelization. Indeed, if you consider single-threaded model than benches would be different.
I just run same benches with a single CPU, it's worse but still 2-4 times faster than naive implementation. Moreover, these benches doesn't not account for the fact that we wouldn't need a lot of calls to Plutus built-in functions (as we do in naive implementation), which I hope will also help to save a lot of CPU budget.

To see to full picture we indeed need implementation in the relevant environment.
BTW, 2 CPUs would already bring speed up to the level outlined in our benchmakrs.